### PR TITLE
Improve performance of heap snapshot.

### LINKED ISSUE #4822: slow performance when taking a heap snapshot
we should batch work so we don't freeze the ui while a snapshot is being taken and the data is being processed.

### PR DESCRIPTION
**Pull request by polina-c**
closes #4822

### CODE REVIEW

**Comment by polina-c** [comment]
Initial benchmark for snapshot: 30 seconds for gallery homepage
Conversion of methods to late final and excluding toLongString from equality operator reduced time to 10 seconds.
Async processing increased time to 20 seconds.

File: packages/devtools_app/lib/src/shared/memory/adapted_heap_data.dart [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import 'package:collection/collection.dart';
 import 'package:vm_service/vm_service.dart';
 
+import '../../../devtools_app.dart';
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/memory/adapted_heap_data.dart:8]: Avoid importing `devtools_app.dart`.

File: packages/devtools_app/lib/src/shared/memory/class_name.dart [hunk 2]
```diff
@@ -102,15 +101,16 @@ class HeapClassName {
...
-  bool get isSentinel => className == 'Sentinel' && library.isEmpty;
+  late final isSentinel = className == 'Sentinel' && library.isEmpty;
 
-  bool get isRoot => className == 'Root' && library.isEmpty;
+  late final isRoot = className == 'Root' && library.isEmpty;
 
   /// Whether a class can hold a reference to an object
   /// without preventing garbage collection.
-  bool get isWeakEntry {
+  late final bool isWeakEntry = () {
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/memory/class_name.dart:113]: See comments on https://github.com/flutter/devtools/pull/5135

**Review by bkonyi — COMMENTED** [review]
I think most of my comments on https://github.com/flutter/devtools/pull/5135 apply here as well.

File: packages/devtools_app/lib/src/shared/memory/adapted_heap_data.dart (hunk 1 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/memory/adapted_heap_data.dart:8]: addressed in https://github.com/flutter/devtools/pull/5135

File: packages/devtools_app/lib/src/shared/memory/class_name.dart (hunk 2 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/memory/class_name.dart:113]: Separated method from initializer.

**Comment by kenzieschmoll** [comment]
20 seconds still seems like a long time. Are there ways we can optimize further?

**Comment by polina-c** [comment]
> 20 seconds still seems like a long time. Are there ways we can optimize further?

Will try in separate PR.

**Pull request merged**